libdrgn/program.c

```diff
@@ -1825,164 +1848,205 @@ static int symbols_search_cb(Dwfl_Module *dwfl_module, void **userdatap,
 							   &elf_sym, &elf_addr,
 							   NULL, NULL, NULL);
-		if (!name || !symbol_match(arg, elf_addr, &elf_sym, name))
+		if (!name || !elf_symbol_match(arg, elf_addr, &elf_sym, name))
 			continue;
 
 		struct drgn_symbol *sym = malloc(sizeof(*sym));
 		if (!sym)
 			return DWARF_CB_ABORT;
 		drgn_symbol_from_elf(name, elf_addr, &elf_sym, sym);
-		if (!symbolp_vector_append(&arg->results, &sym)) {
-			drgn_symbol_destroy(sym);
+		if (elf_symbol_store_match(arg, sym))
 			return DWARF_CB_ABORT;
-		}
 	}
 	return DWARF_CB_OK;
 }
 
-static struct drgn_error *
-symbols_search(struct drgn_program *prog, struct symbols_search_arg *arg,
-	       struct drgn_symbol ***syms_ret, size_t *count_ret)
+struct drgn_error *
+elf_symbols_search(const char *name, uint64_t addr, struct drgn_module *module,
+		   enum drgn_find_symbol_flags flags, void *data,
+		   struct drgn_symbol_result_builder *builder)
 {
-	struct drgn_error *err;
+	Dwfl_Module *dwfl_module = NULL;
+	struct drgn_program *prog = data;
+	struct elf_symbols_search_arg arg = {
+		.name = name,
+		.address = addr,
+		.flags = flags,
+		.err = NULL,
+		.builder = builder,
+	};
 
-	if (!prog->dbinfo) {
-		return drgn_error_create(DRGN_ERROR_MISSING_DEBUG_INFO,
-					 "could not find matching symbols");
-	}
+	if (module)
+		dwfl_module = module->dwfl_module;
 
-	symbolp_vector_init(&arg->results);
+	if (arg.flags & DRGN_FIND_SYM_ADDR) {
+		if (!dwfl_module)
+			dwfl_module = dwfl_addrmodule(prog->dbinfo->dwfl,
+						      arg.address);
+		if (!dwfl_module)
+			return NULL;
+	}
 
-	/*
-	 * When searching for addresses, we can identify the exact module to
-	 * search. Otherwise we need to fall back to an exhaustive search.
-	 */
-	err = NULL;
-	if (arg->flags & SYMBOLS_SEARCH_ADDRESS) {
-		Dwfl_Module *module = dwfl_addrmodule(prog->dbinfo->dwfl,
-						      arg->address);
-		if (module && symbols_search_cb(module, NULL, NULL, 0, arg))
-			err = &drgn_enomem;
+	if ((arg.flags & (DRGN_FIND_SYM_ADDR | DRGN_FIND_SYM_ONE))
+	    == (DRGN_FIND_SYM_ADDR | DRGN_FIND_SYM_ONE)) {
+		GElf_Off offset;
+		GElf_Sym elf_sym;
+		const char *name = dwfl_module_addrinfo(
+			dwfl_module, addr, &offset,
+			&elf_sym, NULL, NULL, NULL);
+		if (!name)
+			return NULL;
+		struct drgn_symbol *sym = malloc(sizeof(*sym));
+		if (!sym)
+			return &drgn_enomem;
+		drgn_symbol_from_elf(name, addr - offset, &elf_sym, sym);
+		arg.err = drgn_symbol_result_builder_add(builder, sym);
+	} else if (dwfl_module) {
+		elf_symbols_search_cb(dwfl_module, NULL, NULL, 0, &arg);
 	} else {
-		if (dwfl_getmodules(prog->dbinfo->dwfl, symbols_search_cb, arg,
-				    0))
-			err = &drgn_enomem;
+		dwfl_getmodules(prog->dbinfo->dwfl, elf_symbols_search_cb, &arg, 0);
 	}
+	return arg.err;
+}
 
+static struct drgn_error *
+drgn_program_symbols_search(struct drgn_program *prog, const char *name,
+			    uint64_t addr, struct drgn_module *module,
+			    enum drgn_find_symbol_flags flags,
+			    struct drgn_symbol_result_builder *ret)
```

Comment:
I think the reason I had it written as an out parameter so that I could include the repetitive error handling logic in this function. And it's a "result builder" rather than an array and length, because I wanted to avoid the extra allocation of a small array for the case where you're looking up a single symbol.

I see why it's a bit unidiomatic though. I'll see if there's a nicer way to do it.